proposals/nnnn-strict-memory-safety.md

```diff
@@ -0,0 +1,536 @@
+# Opt-in Strict Memory Safety Checking
+
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: [Doug Gregor](https://github.com/DougGregor)
+* Review Manager: TBD
+* Status:  **Awaiting review**
+* Vision: [Opt-in Strict Memory Safety Checking (Prospective)](https://github.com/swiftlang/swift-evolution/pull/2581)
+* Implementation:  On main with experimental feature flags `AllowUnsafeAttribute` and `WarnUnsafe`
+* Review: ([pitch](https://forums.swift.org/t/pitch-opt-in-strict-memory-safety-checking/76689))
+
+## Introduction
+
+[Memory safety](https://en.wikipedia.org/wiki/Memory_safety) is a popular topic in programming languages nowadays. Essentially, memory safety is a property that prevents programmer errors from manifesting as [undefined behavior](https://en.wikipedia.org/wiki/Undefined_behavior) at runtime. Undefined behavior effectively breaks the semantic model of a language, with unpredictable results including crashes, data corruption, and otherwise-impossible program states, which directly lead to hard-to-reproduce problems as well as security problems.
+
+Swift provides memory safety with a combination of language affordances and runtime checking. However, Swift also deliberately includes some unsafe constructs, such as the `Unsafe` pointer types in the standard library, language features like `nonisolated(unsafe)`, and interoperability with unsafe languages like C. For most Swift developers, this is a pragmatic solution that appropriate level of memory safety while not getting in the way.
+
+However, there are certain projects, organizations, and code bases that require stronger memory-safety guarantees, such as in security-critical subsystems handling untrusted data or that are executing with elevated privileges in an OS. This proposal introduces opt-in strict memory safety checking to identify those places in Swift code that make use of unsafe language constructs and APIs. Any code written within this strictly-safe subset also works as “normal” Swift and can interoperate with existing Swift code.
```

Comment:
"there are certain projects" reads as euphemistic to me, like you're alluding to top secret work. "some projects" would be a lot less leading.

I think you can set up the logic a little more clearly here. Some projects want stronger safety guarantees. Those projects want to pay closer attention to their unsafe constructs and discourage casual use. This proposal adds a mode to allow projects to opt in to that. Code in this mode is still valid as normal Swift and perfectly compatible.